compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1026,7 +1026,10 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table, posUnpickle
             case APPLIEDtpt =>
               AppliedTypeTree(readTpt(), until(end)(readTpt()))
             case ANDtpt =>
-              AndTypeTree(readTpt(), readTpt())
+              val tpt1 = readTpt()
+              val tpt2 = readTpt()
+              // FIXME: We need to do this instead of "AndType(tpt1, tpt2)" to avoid self-type cyclic reference in tasty_tools
+              untpd.AndTypeTree(tpt1, tpt2).withType(AndType(tpt1.tpe, tpt2.tpe))
```

Comment:
These changes look good, but I wonder whether we should generalize them: Never use & or simplify when dealing with type trees. This looks logically consistent. Have you tried this?